### PR TITLE
Improve CI testing configurations of build products

### LINKED ISSUE #839: Modify build image of AppImage template to use Ubuntu 20.04
According to Github, by Dec 1 2022 they will [formally deprecate the Ubuntu-18.04 Actions runner](https://github.com/actions/runner-images/issues/6002). 

This puts Briefcase in an interesting situation. On the one hand Briefcase as used by end users is (mostly) unaffected - we encourage the use of Docker for building AppImages, and the Ubuntu 18.04 image will continue to exist. 

However, as a project, *we* have a problem. Due to historical issues using Docker in CI, we run our CI in `--no-docker` mode - which means that in December, we'll no longer be able to run CI for Linux. 

**Describe the solution you'd like**

Update the template and documentation to use Ubuntu 20.04 as the base image. 

This is something we've done before ([transitioning from 16.04 to 18.04 as the base image](https://github.com/beeware/briefcase-linux-appimage-template/commit/a172088d46f86b0a95799e8f03bc04149741c765). Interestingly, this was almost 2 years ago to the day, which matches the 2 year LTS release cadence of Ubuntu.

In addition to continued CI support, this has the added benefit that we'll have access to more recent packages in the base image.

**Describe alternatives you've considered**

1. Remove AppImages from our testing setup, but otherwise do nothing. Ubuntu 18.04 AppImage images will continue to work; they'll just be untested.
2. Modify CI to use Docker, but continue to use 18.04 as the base image. 
3. Deprecate the AppImage backend in favour of Flatpak.

**Additional context**

This may have some overlap with #824 - a switch of this magnitude requires some sort of control from a user perspective.

### PR DESCRIPTION
**Pull request by freakboy3742**
Github Actions is scheduled to deprecate the ubuntu-18.04 test runner; this means we'll no longer be able to use `--no-docker` in CI. The use of `--no-docker` in CI has also led to some edge cases slipping through our testing, so this is probably desirable anyway.

Since we need to revisit our CI config anyway, this is also an opportune time to improve CI coverage of other testing platforms. This config change adds iOS, Android, Linux Flatpak, macOS Xcode and Windows VisualStudio tests.

It *doesn't* include a full execution of the final product, so it's not a complete solution to #699.

Fixes #839 
Refs #699

## PR Checklist:

- [x] All new features have been tested
- [x] All new features have been documented
- [x] I have read the **CONTRIBUTING.md** file
- [x] I will abide by the code of conduct

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -131,14 +131,194 @@ jobs:
       run: |
         tox -e py
 
-  verify-desktop:
-    name: Desktop app verification
-    needs: smoke
+  verify-desktop-macOS:
+    name: macOS desktop app verification
+    needs: platform-compat
```
Comment (mhsmith) [inline .github/workflows/ci.yml:136]: Is it necessary to have 5 configurations which have so much in common? Could the differences (e.g. cache directories, `--adhoc-sign` argument) and matrix limitations (e.g. building iOS on macOS only) have been dealt with using `include` variables and `if`?

Comment (freakboy3742) [reply .github/workflows/ci.yml:136]: I definitely agree this is... verbose. The original version of this stage was a matrix build; however, adding iOS and Android as another axis, while only building Toga got prohibitively complex. However, thinking about it again, it might be easier to do this as a single OS x project matrix, and then exclude iOS and Android based on a single axis of the matrix. I'll have a fiddle and see what I can do.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -131,14 +131,194 @@ jobs:
...
+          ~/.cache/pip
+    - uses: actions/checkout@v3
+    - name: Set up Python
+      uses: actions/setup-python@v3.1.2
+      with:
+        python-version: ${{ env.python_version }}
+    - name: Install system dependencies
+      run: |
+        sudo apt-get update -y
+        sudo apt-get install -y python3-gi python3-gi-cairo gir1.2-gtk-3.0 python3-dev libgirepository1.0-dev libcairo2-dev pkg-config flatpak flatpak-builder
```
[the remark this answers is not in the document]
Comment (freakboy3742) [reply .github/workflows/ci.yml:251]: I'm fairly sure we do - these packages aren't in the [list of apt packages in the 22.04 runner config](https://github.com/actions/runner-images/blob/main/images/linux/Ubuntu2204-Readme.md#installed-apt-packages). However, I'll experiment and see.

[the remark this answers is not in the document]
Comment (freakboy3742) [reply .github/workflows/ci.yml]: Oh - wait - now I see what you're saying: we don't need these because we're building *in* docker. You're completely correct - we only need the flatpak tools.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (freakboy3742) [reply .github/workflows/ci.yml:136]: Ok - I've been able to clean this up quite a bit. It turns out it was the use of `tox` that was making the process complex; once I dropped Tox, the matrix and step definitions actually ended up quite elegant.

**Review by mhsmith — APPROVED** [review]
